lib/luacrypto_aead.c

```diff
@@ -0,0 +1,258 @@
+}
+
+/***
+* Gets the required initialization vector (IV) size for the AEAD transform.
+* @function crypto_aead:ivsize
+* @treturn integer The IV size in bytes.
+*/
+static int luacrypto_aead_ivsize(lua_State *L) {
+	struct crypto_aead *tfm = luacrypto_aead_check(L, 1);
+	lua_pushinteger(L, crypto_aead_ivsize(tfm));
+	return 1;
+}
+
+/***
+* Gets the current authentication tag size for the AEAD transform.
+* This is the value set by `setauthsize` or the algorithm's default.
+* @function crypto_aead:authsize
+* @treturn integer The authentication tag size in bytes.
+*/
+static int luacrypto_aead_authsize(lua_State *L) {
+	struct crypto_aead *tfm = luacrypto_aead_check(L, 1);
+	lua_pushinteger(L, crypto_aead_authsize(tfm));
+	return 1;
+}
+
+/* Auxiliary functions to define luacrypto_aead_encrypt and luacrypto_aead_decrypt */
+
+static inline size_t luacrypto_aead_get_encrypt_buffer_needed(lua_State *L, size_t combined_len, unsigned int authsize_val, size_t crypt_len) {
+	return combined_len + authsize_val;
+}
+
+static inline size_t luacrypto_aead_get_decrypt_buffer_needed(lua_State *L, size_t combined_len, unsigned int authsize_val, size_t crypt_len) {
+	luaL_argcheck(L, crypt_len >= authsize_val, 3, "input data (ciphertext+tag) too short for tag");
+	return combined_len;
+}
+
+static inline int luacrypto_aead_common_crypt(lua_State *L, int (*crypt_func)(struct aead_request *req), size_t (*get_buffer_needed_func)(lua_State *L, size_t combined_len, unsigned int authsize_val, size_t crypt_len), int result_adjust_factor)
+{
+	struct crypto_aead *tfm = luacrypto_aead_check(L, 1);
+	size_t iv_len, combined_len;
+	const char *iv = luaL_checklstring(L, 2, &iv_len);
+	const char *combined = luaL_checklstring(L, 3, &combined_len);
+	lua_Integer aad_l = luaL_checkinteger(L, 4);
+
+	unsigned int expected_ivlen = crypto_aead_ivsize(tfm);
+	luaL_argcheck(L, iv_len == expected_ivlen, 2, "incorrect IV length");
+	u8 *iv_data = lunatik_checkalloc(L, iv_len);
+	memcpy(iv_data, iv, iv_len);
+
+	luaL_argcheck(L, aad_l >= 0 && (size_t)aad_l <= combined_len, 4, "AAD length out of bounds");
+	size_t aad_len = (size_t)aad_l;
+
+	unsigned int authsize_val = crypto_aead_authsize(tfm);
+
+	luaL_Buffer b;
+	luaL_buffinit(L, &b);
+	char *buf;
+
+	size_t crypt_len = combined_len - aad_len; /* plaintext_len for encrypt, ciphertext_with_tag_len for decrypt */
+
+	size_t total_buffer_needed = get_buffer_needed_func(L, combined_len, authsize_val, crypt_len);
+	buf = luaL_prepbuffsize(&b, total_buffer_needed);
+
+	memcpy(buf, combined, combined_len);
+	luaL_addsize(&b, combined_len);
+
+	struct scatterlist sg_work;
+	sg_init_one(&sg_work, buf, total_buffer_needed);
+
+	gfp_t gfp = lunatik_gfp(lunatik_toruntime(L));
+	struct aead_request *req = aead_request_alloc(tfm, gfp);
+	if (!req) {
+		lunatik_free(iv_data);
+		luaL_error(L, "Failed to allocate request");
+	}
+	aead_request_set_ad(req, aad_len);
+	aead_request_set_crypt(req, &sg_work, &sg_work, crypt_len, iv_data);
```

Comment:
you only use iv_data here, right? you can allocate it just before.. btw, why do you need to memcpy this? why can't you just use iv directly here?